### ISSUE TITLE
Delay entity insert may fail with Merge

### ISSUE THREAD

**Issue by simopala** [nhibernate/nhibernate-core#2608]
We are having problems upgrading to new version 5.3.0 due to a breaking change on delay entity insert on persist: #1754

I read the discusson and agreed that Persist should delay entity insert, but Save does it immediatly.

Is it correct that Merge methos doesn't create identity immediatly?
I think that Merge method should work like Save.

For this reason, if I create a OneToOne relation between two objects with the same primary key:
- Merge gives "Specific cast is not valid"
- Save works correctly

Example:

```
public TableAMap()
        {
            Schema("[dbo]");
            Table("TEST_TABLE_A");
            Not.LazyLoad();

            Id(x => x.Id).Column("ID");

            Map(x => x.Name).Column("NAME").Length(100).Not.Nullable();
            Map(x => x.Description).Column("DESCRIPTION").Length(1000).Nullable();
            Map(x => x.CreationDate).Column("CREATION_DATE").Not.Nullable();

            HasOne(x => x.TableB).Not.LazyLoad().Cascade.All();
        }
```

```
public TableBMap()
        {
            Schema("[dbo]");
            Table("TEST_TABLE_B");
            Not.LazyLoad();

            Id(x => x.Id).Column("ID").GeneratedBy.Foreign("TableA");
            Map(x => x.Color).Column("COLOR").Length(100).Nullable();

            HasOne(x => x.TableA).Constrained();
        }
```

```
using (var session = sessionFactory.OpenSession())
            {
                using (var transaction = session.BeginTransaction())
                {
                    var tableA = new TableA {Name = "test", Description = "it works?", CreationDate = DateTime.UtcNow};
                    tableA.TableB = new TableB { TableA = tableA, Color = "red" };

                    session.Save(tableA);
                    //session.Merge(tableA);

                    transaction.Commit();

                    var test = session.Query<TableA>().FirstOrDefault(x => x.Id == tableA.Id);
                }
            }
```

**Comment by bahusoid** [comment]
> Merge gives "Specific cast is not valid"

Exception stacktrace? From you description it's actually not clear that it's somehow related to delayed entity insert...

**Comment by fredericDelaporte** [comment]
Adjusting one of our tests reproduces it:
```
System.InvalidCastException : Le cast spécifié n'est pas valide.
   à (Object , Object )
   à NHibernate.Bytecode.Lightweight.AccessOptimizer.SetSpecializedPropertyValue(Object target, Object value) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Bytecode\Lightweight\AccessOptimizer.cs:ligne 76
   à NHibernate.Bytecode.AccessOptimizerExtensions.SetSpecializedPropertyValue(IAccessOptimizer optimizer, Object target, Object value) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Bytecode\IAccessOptimizer.cs:ligne 68
   à NHibernate.Tuple.Entity.PocoEntityTuplizer.SetIdentifierPropertyValue(Object entity, Object value) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Tuple\Entity\PocoEntityTuplizer.cs:ligne 375
   à NHibernate.Tuple.Entity.AbstractEntityTuplizer.SetIdentifier(Object entity, Object id) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Tuple\Entity\AbstractEntityTuplizer.cs:ligne 160
   à NHibernate.Persister.Entity.AbstractEntityPersister.SetIdentifier(Object obj, Object id) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Persister\Entity\AbstractEntityPersister.cs:ligne 4464
   à NHibernate.Event.Default.AbstractSaveEventListener.PerformSave(Object entity, Object id, IEntityPersister persister, Boolean useIdentityColumn, Object anything, IEventSource source, Boolean requiresImmediateIdAccess) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\AbstractSaveEventListener.cs:ligne 176
   à NHibernate.Event.Default.AbstractSaveEventListener.SaveWithGeneratedId(Object entity, String entityName, Object anything, IEventSource source, Boolean requiresImmediateIdAccess) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\AbstractSaveEventListener.cs:ligne 129
   à NHibernate.Event.Default.DefaultMergeEventListener.SaveTransientEntity(Object entity, String entityName, Object requestedId, IEventSource source, IDictionary copyCache) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\DefaultMergeEventListener.cs:ligne 293
   à NHibernate.Event.Default.DefaultMergeEventListener.MergeTransientEntity(Object entity, String entityName, Object requestedId, IEventSource source, IDictionary copyCache) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\DefaultMergeEventListener.cs:ligne 244
   à NHibernate.Event.Default.DefaultMergeEventListener.EntityIsTransient(MergeEvent event, IDictionary copyCache) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\DefaultMergeEventListener.cs:ligne 214
   à NHibernate.Event.Default.DefaultMergeEventListener.OnMerge(MergeEvent event, IDictionary copiedAlready) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\DefaultMergeEventListener.cs:ligne 174
   à NHibernate.Impl.SessionImpl.FireMerge(IDictionary copiedAlready, MergeEvent event) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Impl\SessionImpl.cs:ligne 2199
   à NHibernate.Impl.SessionImpl.Merge(String entityName, Object obj, IDictionary copiedAlready) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Impl\SessionImpl.cs:ligne 835
   à NHibernate.Engine.CascadingAction.MergeCascadingAction.Cascade(IEventSource session, Object child, String entityName, Object anything, Boolean isCascadeDeleteEnabled) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Engine\CascadingAction.cs:ligne 270
   à NHibernate.Engine.Cascade.CascadeToOne(Object parent, Object child, IType type, CascadeStyle style, Object anything, Boolean isCascadeDeleteEnabled) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Engine\Cascade.cs:ligne 275
   à NHibernate.Engine.Cascade.CascadeAssociation(Object parent, Object child, IType type, CascadeStyle style, Object anything, Boolean isCascadeDeleteEnabled) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Engine\Cascade.cs:ligne 240
   à NHibernate.Engine.Cascade.CascadeProperty(Object parent, Object child, IType type, CascadeStyle style, String propertyName, Object anything, Boolean isCascadeDeleteEnabled) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Engine\Cascade.cs:ligne 150
   à NHibernate.Engine.Cascade.CascadeOn(IEntityPersister persister, Object parent, Object anything) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Engine\Cascade.cs:ligne 128
   à NHibernate.Event.Default.AbstractSaveEventListener.CascadeAfterSave(IEventSource source, IEntityPersister persister, Object entity, Object anything) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\AbstractSaveEventListener.cs:ligne 376
   à NHibernate.Event.Default.DefaultMergeEventListener.MergeTransientEntity(Object entity, String entityName, Object requestedId, IEventSource source, IDictionary copyCache) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\DefaultMergeEventListener.cs:ligne 280
   à NHibernate.Event.Default.DefaultMergeEventListener.EntityIsTransient(MergeEvent event, IDictionary copyCache) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\DefaultMergeEventListener.cs:ligne 214
   à NHibernate.Event.Default.DefaultMergeEventListener.OnMerge(MergeEvent event, IDictionary copiedAlready) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\DefaultMergeEventListener.cs:ligne 174
   à NHibernate.Event.Default.DefaultMergeEventListener.OnMerge(MergeEvent event) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Event\Default\DefaultMergeEventListener.cs:ligne 41
   à NHibernate.Impl.SessionImpl.FireMerge(MergeEvent event) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Impl\SessionImpl.cs:ligne 2186
   à NHibernate.Impl.SessionImpl.Merge(String entityName, Object obj) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Impl\SessionImpl.cs:ligne 909
   à NHibernate.Impl.SessionImpl.Merge(Object obj) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Impl\SessionImpl.cs:ligne 925
   à NHibernate.Impl.SessionImpl.Merge[T](T entity) dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate\Impl\SessionImpl.cs:ligne 915
   à NHibernate.Test.Operations.MergeFixture.MergeBidiPrimayKeyOneToOne() dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate.Test\Operations\MergeFixture.cs:ligne 132
```
Adjustments: switch `Person` id to `native`, fix `PersonalDetails` id declaration as `foreign` instead of relying on luck to have it the same than the parent, remove the `Ignore` from the `MergeBidiPrimayKeyOneToOne` test, change its `Persist` to `Merge`.

**Comment by fredericDelaporte** [comment]
Cherry-picking my adjustments to a 5.2.x branch, no failure. This is a 5.3.x regression.

**Comment by simopala** [comment]
Sorry for the delay.
Thanks @fredericDelaporte, my stack trace is similar to yours.
I will wait upgrading to the new versions till a new version will be released.

**Comment by fredericDelaporte** [comment]
Fixed by #2609
